Finished Sell Land module + test

lisv opened the pull request:
User Stories: 
- Implement strategy for when and how to sell land
- Implement additional strategy for selling land when a reaction has been received
- Write tests for the features you're implementing this sprint

We only want to sell buildings that are not part of housing, as we do not need those. We first try to sell these multipolygons to the facilities stakeholder, for a maximum of three times, for which we reduce the selling price each time. 

If they still did not buy it we try selling it to other stakeholders in the same way.

lisv committed:
Add quality and sound indicators in knowledge base, insert land and zone percepts (besides the lists),  adjust goals

lisv committed:
Add goal adopting for selling land when the liveability factor in the zone the multipolygon is in, is too low for our standards

lisv committed:
Revert "Add goal adopting for selling land when the liveability factor in the zone the multipolygon is in, is too low for our standards"

This reverts commit deb419054affb7bc967abb5552c4c231f9b145f6.

lisv committed:
Adjust quality and sound indicator definitions

lisv committed:
Attempt to fix failing build

lisv committed:
Comment out things that are probably causing the failing build

lisv committed:
Merge branch 'dev' into selling-land

# Conflicts:
#	src/nl/codefox/contextvh/knowledgebase.pl
#	src/nl/codefox/contextvh/tygronEvents.mod2g
#	src/nl/codefox/contextvh/tygronGoals.mod2g
#	src/nl/codefox/contextvh/tygronInit.mod2g

lisv committed:
Adjust the way we mark land sold

lisv committed:
Add the forgotten dynamic tag

lisv committed:
Set quality and sound indicators in knowledge

lisv committed:
Add goal adopting for selling buildings that we do not need, remove unused variable myBuildingList, adjust ownedBuilding predicate to avoid duplicate code

lisv committed:
Remove white space

lisv committed:
Merge branch 'dev' into selling-land

# Conflicts:
#	src/nl/codefox/contextvh/knowledgebase.pl

lisv committed:
Add goals as knowledge variables

Commit:
changed predicate to fix Travis

The previous predicate was looking for categories which could not exist in this form, I also partially refactored the line for a cleaner overview. NotHousingMember comment should be improved.

lisv committed:
Merge branch 'dev' into selling-land

# Conflicts:
#	src/nl/codefox/contextvh/knowledgebase.pl

lisv committed:
Merge branch 'dev' into selling-land

lisv committed:
Remove unnessecary white spaces

lisv committed:
Adapt name of notHousingMember to be more clear about what it is, fix building believe in this branch to match building/10

lisv committed:
Add extra comments in sellLand module

lisv committed:
Delete offeredLand(MultiPoly,Counter) from belief base when we have sold the building/land

lisv committed:
Try to sell to services stakeholder first, stop when we have tried three times. Set price at 400 per m2 as a starting price.

lisv committed:
Try more stakeholders after we tried facilities to sell land to

lisv committed:
Fix variable mistake

lisv committed:
Merge branch 'dev' into selling-land

lisv committed:
Sell all buildings that are of no use to us

MatthijsJWH commented:
This module can (at least partially) and should be tested before it is merged in my opinion. Or can you give me a explanation as to why it can't be tested?

lisv commented:
Okay, I'm working on it

lisv committed:
Add sellLand module test

lisv commented:
Passing test has been added.

shdshdw commented on src/nl/codefox/contextvh/knowledgebase.pl:
```diff
@@ -23,50 +23,70 @@
 	MultiPoly\=multipolygon('MULTIPOLYGON EMPTY').
+	
+% categories that are a member of housing
+housingMember(Categories) :- member('LUXE',Categories); member('NORMAL',Categories); member('SOCIAL',Categories); member('OTHER',Categories).
```
I think housingMember is a somewhat confusing name for this predicate.

shdshdw commented on src/nl/codefox/contextvh/sellLand.mod2g:
```diff
@@ -5,15 +5,33 @@ exit=always.
+		insert(doneSelling(MultiPoly)) + exit-module.
+	
+	% try another buyer when we have tried it three times with facilities
+	if bel(offeredLand(MultiPoly,BuyerID,Counter),facilitiesID(BuyerID),Counter=3,
```
I believe it is better to handle this logic to adopt the goal to sell the correct land instead of checking that in this module. Mostly because it will then only execute the module when there really is a need to sell land.

shdshdw commented:
When I run the code, there are no single occurrences where offeredLand(MultiPoly, Counter) is inserted with Counter > 1,and because there is no request handling for sell land I would like to know when does this counter actually become 2 or higher?
Edit: I ran this while Marco acted as the facilities stakeholder. After declining a sell land the counter does not get updated. 
On a sidenote: why do we actually sell about ALL our land?

MHMCD commented:
So we decided to look what is happening in Tygron when the bot is running, we found quite some interesting interactions with the environment. Dare I say so, Terrifying.
As you can see in this beautiful picture:
https://i.gyazo.com/2d51b23706c808387d1b7f37956227da.png
We are currently selling all the gardens around our owned buildings, this is not what I meant with selling certain facilities. Now I understand that you did not take into account that these would be brought up to sell but I would expect of you to at least check in the tygron engine what is happening when your bot is running with your new code, you told us that your tygron engine is not working, then ask us to run it for you so we can give you feedback on what your bot is doing. The counter does not work at the moment, you are selling way too much ground and you believe the code is ready to be merged into the development branch? Could you please at least look at what your bot is doing before you try to pull request your branch.

lisv committed:
Merge branch 'dev' into selling-land

lisv committed:
Add sell land request handling using a counter, adapt sellLand, currently only sells to facilities

Commit:
repair selling land

returned some predicates to their old state, added building to OurLand because some multiPolygons aren't saved as land but owned by us as a building.
NOTE: AT this moment the bot does sell the housing once but does not try to sell it again when declined, the bot also sells roads which is not really what we planned for, should be fixed soon.

Commit:
Selling land fixed

Selling only the building we have to sell up to 4 times and no random sells of intersections/roads.

Commit:
Merge branch 'dev' into selling-land

Conflicts:
	src/nl/codefox/contextvh/knowledgebase.pl

lisv committed:
Add sellTo predicate which gives us information about who we are trying to sell land to and use it in the module, needs to be tested

lisv committed:
Add 'BRIDGE' to housingBuildings as we do not want to sell bridges

lisv committed:
Add functionality to demolish a building once it has been tried to sell before but did not succeed

lisv committed:
Fix mistake so that the agent will sell land to other stakeholder

lisv committed:
Fix code so that the agent will sell to the right buyer and not always to Facilities

lisv commented:
You were right, this branch was not ready to be pulled into the developing branch.

However, I have rewritten it and manually tested it (with Marco) and now it works correctly:
- The sellLand module is tested
- We only sell buildings that are not part of housing now
- The counter works, using the requestsmodule
- The code has been manually tested

Please review it again.

shdshdw commented:
Resolve the conflicts please.

lisv committed:
Merge branch 'dev' into selling-land

# Conflicts:
#	lib/tygron
#	src/nl/codefox/contextvh/knowledgebase.pl
#	src/nl/codefox/contextvh/tygronGoals.mod2g

lisv commented:
Conflicts are currently resolved

shdshdw commented on src/nl/codefox/contextvh/knowledgebase.pl:
```diff
@@ -52,31 +50,57 @@
+	(IndicatorName=='Ruimtelijke kwaliteit'), indicator(IndicatorID,Value,Target,ZoneLink),
+	member(zone_link(ZoneID,IndicatorID,Value,Target),ZoneLink).
+% indicator of the sound
+soundIndicator(Value,Target,ZoneID) :-
```
Although I think it is good to use predicates like this instead of hard-coding indicatorId's, I think this is a change that is outside of the scope of this PR. Please consider this in next PR's.

lisv commented:
Also, a time out should also be implemented so that the agent will move on if it does not receive a reaction back from a stakeholder but this is too big of a task for now. We should do this soon in a different branch.

shdshdw commented on src/nl/codefox/contextvh/tygronGoals.mod2g:
```diff
@@ -4,6 +4,17 @@ order = linearall.
+	
+	% If we want to sell a building but failed to do it before, we demolish it.
+	if bel(building(BuildingID,Name,OwnerID,_,Categories,FunctionID,_,MultiPoly,_,_),ourID(OwnerID),not(housingBuildings(Categories)),
+		failedToSell(MultiPoly))
```
As this functionality has not yet been triggered, please check if this works correctly. Because this functionality does not break other code I agree to do so after the merge.

shdshdw commented:
Approved, but please extend this after the merge.

MatthijsJWH commented:
I approve, nicely done adding the test

MHMCD commented:
I approve, we have done multiple tests and I believe we ruled out the bugs we were experiencing earlier.

JoshCode commented on lib/tygron:
```diff
@@ -1 +1 @@
-Subproject commit 41169d27e37352fc367d4bdfbfa9cf7e3480a1f6
```
This is the wrong submodule commit

lisv replied:
Fixed

lisv committed:
Make submodule refer to the correct commit again

JoshCode commented:
Looks ready to be merged now.

JoshCode commented:
Travis builds should be passing, it's a problem with Travis, not our bot. Merging.
